### PR TITLE
[VAN-124119][VAN-124120][VAN-124123] Snapshot sample application

### PR DESCRIPTION
**Pull request by Shmalikov**

### CODE REVIEW

File: time-machine/README.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
+# Time Machine Sample Application
+
+## Overview
+
+Organizations frequently report HR metrics to external stakeholders and use them for internal scorecarding.
+However, subsequent corrections to historical data can introduce discrepancies between the initially reported values and
+the current ones.
+
+This project addresses the challenge of maintaining data integrity for accurate historical reporting and analysis.
```
Comment (hareskoug) [inline time-machine/README.md:9]:
```suggestion
This sample application addresses the challenge of maintaining data integrity for accurate historical reporting and analysis.
```

File: time-machine/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    - Use the Data Query API to download the current data state and save it in a temporary CSV file.
+    - An aggregate query stored in a JSON file is required for this step. A sample query is located in
+      [./queries/gender-resignation-joblevel.json](queries%2Fgender-resignation-joblevel.json).
+
+2. Uploading this csv file using Data Upload API:
+    - Use the Data Upload API to upload the CSV file.
+    - The Visier platform maps this data to a pre-created Overlay, which will be used to access data as it was at a
+      specific point in time.
+
+To process the uploaded file, the Visier Platform requires the creation of a Source, Overlay, and Mapping.
```
Comment (hareskoug) [inline time-machine/README.md:27]: Should this be moved to prerequisites? It's done only once before the whole thing can run

File: time-machine/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+Here are the instructions:
+
+1. [Source creation](https://docs.visier.com/developer/Studio/data/sources/source-create.htm)
+2. [Overlay creation](https://docs.visier.com/developer/Analytic%20Model/analytic-objects/overlays/overlays-configure.htm)
+3. [Mapping from Source to Overlay](https://docs.visier.com/developer/Studio/data/mappings/mapping-add.htm)
+
+Instead of manually creating the source, overlay, and mapping, you can import these pre-configured components from the
+provided files:
+
+1. [./import/WFF_d3m_source.zip](import%2FWFF_d3m_source.zip)- to import source.
```
Comment (hareskoug) [inline time-machine/README.md:37]: Will this import work on tenants named something other than `WFF_d3m`?

File: time-machine/README.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+```json
+"timeIntervals": {
+"dynamicDateFrom": "COMPLETE_PERIOD",
+"intervalPeriodType": "MONTH",
+"intervalCount": 1
+}
+```
+
+When run every month, this sample will upload the state of the data to the pre-created Overlay.
+The Overlay could be used to check what data was at a specific point in time.
```
Comment (hareskoug) [inline time-machine/README.md:60]:
```suggestion
The Overlay retains all the uploaded snapshots which will enable historical analysis of the data as it was known at the time.
```

File: time-machine/main.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    api_client = ApiClient(config, 'Accept', 'text/csv')
+    query_api = DataQueryApi(api_client)
+
+    logger.info(f"Executing query: {query_file_path}")  # Combined messages
+    response = query_api.aggregate_without_preload_content(query_dto)
+    if not 200 <= response.status <= 299:
+        raise Exception(f"API request failed with status code {response.status}: {response.data.decode()}")
+
+    os.makedirs(os.path.dirname(data_file_path), exist_ok=True)
```
Comment (hareskoug) [inline time-machine/main.py:88]: Consider tracking whether you're actually creating a new directory and if so, delete the dir too before existing

File: time-machine/main.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        data = f.read()
+
+    filename = os.path.basename(data_file_path)
+    logger.info(f"Uploading data: {filename}")  # Kept for visibility
+    upload_api.v1_data_upload_files_filename_put(filename=filename, body=data)
+
+
+def main():
+    try:
+        logger.info("Time Machine started.")
```
Comment (hareskoug) [inline time-machine/main.py:114]: Not sure I like calling this sample app Time Machine. It doesn't allow traveling in time. I'd like to reserve that name for when we're providing a sample that allows you to change the query API to run off a historical DV. That would be more like a time machine. This app is more like snapshot recording. Check with PdM if they have an alternative name they're ok with

File: time-machine/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,4 @@
+python-dotenv>=1.0.1
+visier.sdk.api.data-in>=22222222.99201.1418
```
Comment (hareskoug) [inline time-machine/requirements.txt:2]: I'm still all giddy when I see this. 🙂

File: time-machine/README.md (hunk 2 above)
Comment (Shmalikov) [reply time-machine/README.md:27]: Moved.

File: time-machine/main.py (hunk 6 above)
Comment (Shmalikov) [reply time-machine/main.py]: Wayne agreed to call it Snapshot Recorder, and if there's a better name we'll rename it.

File: time-machine/requirements.txt (hunk 7 above)
Comment (Shmalikov) [reply time-machine/requirements.txt]: Yeah, it is the best part.

File: time-machine/README.md (hunk 3 above)
Comment (Shmalikov) [reply time-machine/README.md]: It worked, but there should be manually assigned Data Category and Tenant Settings.
I changed archives to have 3 parts:
1. source
2. data_category, tenant_settings
3. overlay and mapping

File: snapshot-recorder/main.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        os.rmdir(temp_dir)
+        logger.info(f"Temporary directory {temp_dir} removed.")
+
+
+def main() -> None:
+    try:
+        logger.info("Snapshot recorder started.")
+
+        args = parse_args()
+        load_dotenv(dotenv_path='.env.time-machine', override=True)
```
Comment (hareskoug) [inline snapshot-recorder/main.py:160]: Nit: rename env file now that it's no longer called time-machine

File: time-machine/main.py (hunk 5 above)
Comment (Shmalikov) [reply time-machine/main.py]: Added removing dir

File: snapshot-recorder/main.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+        download_data(api_config, args.query_file_path, temp_data_file_path)
+
+        upload_data(api_config, temp_data_file_path)
+
+        clean_temp_dir(temp_dir, temp_dir_created, temp_data_file_path)
+
+        logger.info("Snapshot recorder finished.")
+    except Exception as e:
+        logger.error(f"An unexpected error occurred: {e}")
```
Comment (hareskoug) [inline snapshot-recorder/main.py:176]: Clean up in the event of failure

File: snapshot-recorder/main.py (hunk 8 above)
Comment (Shmalikov) [reply snapshot-recorder/main.py]: Renamed to snapshot-recorder

File: snapshot-recorder/main.py (hunk 9 above)
Comment (Shmalikov) [reply snapshot-recorder/main.py:176]: Decided to keep files to be able to upload it later.

**Review by hareskoug — APPROVED** [review]

**Pull request merged**